File: src/rescore/rescore.py
#!/usr/bin/env python

# TODO Rename symbol_set -> symbols_set
# TODO Rename symbols_type -> org_symbol_scores

import argparse
import os
import math
import requests
import json
import sys
import time

import numpy as np
from Perceptron import Perceptron

from statistics import get_file_stats
from utility import get_all_filenames, shuffle

def get_dataset_from_logs(logdir):
    '''
    Returns X, y, symbol_set
    X is a list of lists. Each list contains symbols hit by an email
    y is a list of actual email class. Yi = 1 ith email is spam, 0 otherwise
    symbol_set is a list of symbols
    '''
    
    filenames = get_all_filenames(logdir)

    X = []
    y = []
    
    for file in filenames:
        with open(file, 'r') as f:
            for line in f:
                X.append(line.split()[4:])
                y.append(0 if line.split()[1] == "HAM" else 1)

    symbol_set = get_all_symbols(X)

    X, y = make_perceptron_input(X, y, symbol_set)

    X = np.array(X)
    y = np.array(y)

    return X, y, symbol_set


def get_all_symbols(dataset):

    symbol_set = set()

    for row in dataset:
        map(lambda x : symbol_set.add(x), row)

    return tuple(symbol_set)


def make_perceptron_input(X, y, symbol_set):
    '''
    Returns a list of lists containing 1s and 0s. 
    X(i, j) = 1 if symbol_set(j) is present in ith log line, 0 otherwise
    '''

    X_new = []

    for row in X:
        X_new.append([1 if symbol in row else 0 for symbol in symbol_set])

    return np.array(X_new), np.array(y)


def file_exists(filepath):

    return os.path.isfile(filepath)


def get_symbols_type_from_file(filepath):

    cache = {}
    
    with open(filepath, 'r') as f:
        cache = json.load(f)

    return cache


def get_all_symbols_type(address="localhost", port=11334, endpoint="symbols"):

    url = "http://{}:{}/{}".format(address, port, endpoint)

    response = json.loads(requests.get(url).text)

    symbols_type = {}

    for group in response:
        group = group["rules"]
        for rule in group:
            symbols_type[rule['symbol']] = rule['weight']

    return symbols_type


def write_symbols_cache(symbols_type, filepath="symbols_type.cache"):

    with open(filepath, "w") as f:
        json.dump(symbols_type, f)

    return


def get_symbols_type_from_file(filepath):

    symbols_type = {}
    
    with open(filepath, 'r') as f:
        symbols_type = json.load(f)
    
    return symbols_type


def filter_symbols_type(symbols_set, symbols_type):

    filtered_symbols_type = {}

    for symbol in symbols_set:
        filtered_symbols_type[symbol] = 0
        if symbol in symbols_type:
            filtered_symbols_type[symbol] = symbols_type[symbol]
    
    return filtered_symbols_type
    

def get_symbols_type(symbol_set, symbols_type_cache_file="symbols_type.cache"):

    symbols_type = {}
    
    if file_exists(symbols_type_cache_file):
        symbols_type = get_symbols_type_from_file(symbols_type_cache_file)

    else:
        symbols_type = get_all_symbols_type()
        write_symbols_cache(symbols_type)

    symbols_type = filter_symbols_type(symbol_set, symbols_type) # Removes symbols not in symbol set

    return symbols_type


def rescore_weights(X, y, symbols_tuple, symbols_type, threshold, decay, epoch, l_rate):
    '''
    Returns a tuple of (symbol, score) after training perceptron
    '''

    n_samples, n_feaures = len(X), len(X[0])

    perceptron = Perceptron(symbols_tuple=symbols_tuple,
                            n_epoch=epoch,
                            l_rate=l_rate,
                            threshold=threshold,
                            decay=decay,
                            symbols_type=symbols_type)

    weights = perceptron.rescore_weights(X, y)

    return weights
    

def split_dataset(X, y, percent):
    '''
    Splits dataset into (percent) % and (100 - percent) %
    Returns X1, y1, X2, y2
    '''

    split_index = len(X) * percent / 100

    X1 = X[ : split_index]
    y1 = y[ : split_index]

    X2 = X[split_index + 1 : ]
    y2 = y[split_index + 1 : ]

    return X1, y1, X2, y2


def print_new_scores(output_file, symbol_set, symbols_type, new_scores):

    score_output_format = "{:<35} {:<13} {:<10}"

    print >>output_file, score_output_format.format("SYMBOL", "OLD SCORE", "NEW SCORE")

    for i in range(len(symbol_set)):
        print >>output_file, score_output_format.format(symbol_set[i], symbols_type[symbol_set[i]], new_scores[i])

        
def eval_email_score(record, new_scores):

    score = 0
    
    for i in range(len(record)):
        if record[i] == 1:
            score = score + new_scores[i]

    return score


def make_log_for_stats(X, y, scores):

    logs = []

    for x, y_val in zip(X, y):
        logs.append("dummy_file" + " " +
                    ("HAM" if y_val == 0 else "SPAM") + " " +
                    str(eval_email_score(x, scores)) + " " +
                    "DUMMY_SYM" +
                    "\n")

    return logs


def find_best_spam_threshold(X_cv, y_cv, spam_thresholds, new_scores):

    logs = make_log_for_stats(X_cv, y_cv, new_scores)

    max_accuracy = -1
    max_accuracy_threshold = -1
    ms = 0
    
    for threshold in spam_thresholds:
        stats = get_file_stats(logs, threshold)
        accuracy = stats[-1]
        if(accuracy > max_accuracy):
            max_accuracy = accuracy
            max_accuracy_threshold = threshold
            ms = stats    

    return max_accuracy_threshold


def fscore(stats):

    fp = stats.false_positive_rate * stats.no_of_ham / 100
    fn = stats.false_negative_rate * stats.no_of_spam / 100
    
    f_score = 2 * stats.true_positives / float(2 * stats.true_positives + fp + fn)

    return f_score


def print_stats(X, y, scores, threshold):

    stats = get_file_stats(make_log_for_stats(X, y, scores), threshold)
    f_score = fscore(stats)

    print
    print "Statistics at threshold {}".format(threshold)
    print "Accuracy: {} %".format(str(round(stats.overall_accuracy, 2)))
    print "F score: {}".format(fscore(stats))
    print "False positive rate: {}".format(stats.false_positive_rate)
    print "False negative rate: {}".format(stats.false_negative_rate)
    print
    
    
def main():

    start_time = time.time()
    
    epoch = 100
    l_rate = 0.001
    threshold = 15
    output = sys.stdout
    decay = 1
    
    arg_parser = argparse.ArgumentParser()
    arg_parser.add_argument("-l", "--logdir",
                            help="path to log directory")
    arg_parser.add_argument("-e", "--epoch",
                            help="no of epochs [Default: 100]",
                            type=int)
    arg_parser.add_argument("-r", "--lrate",
                            help="Learning rate of perceptron [Default: 0.001]",
                            type=float)
    arg_parser.add_argument("-t", "--threshold",
                            help="threshold value [Default: 15]",
                            type=float)
    arg_parser.add_argument("-o", "--output",
                            help="Write new scores to file")
    arg_parser.add_argument("-d", "--decay",
                            help="Weight decay [Default: 1]",
                            type=float)
    
    args = arg_parser.parse_args()

    if not args.logdir:
        arg_parser.error('specify log directory')
        return

    if args.epoch:
        epoch = args.epoch

    if args.lrate:
        l_rate = args.lrate

    if args.threshold:
        threshold = args.threshold

    if args.output:
        output = open(args.output, 'w')

    if args.decay:
        decay = args.decay
        
    X, y, symbol_set = get_dataset_from_logs(logdir=args.logdir)

    X, y = shuffle(X, y)
    
    symbols_type = get_symbols_type(symbol_set)

    # Split data into 60 : 20 : 20 (Train : Cross-validation : Test)
    X_train, y_train, X_test, y_test = split_dataset(X, y, 60)

    X_cv, y_cv, X_test, y_test = split_dataset(X_test, y_test, 50) 

    old_scores = []
    for sym in symbol_set:
       old_scores.append(symbols_type[sym])

    new_scores = rescore_weights(X=X_train,
                                 y=y_train,
                                 epoch=epoch,
                                 l_rate=l_rate,
                                 decay=decay,
                                 threshold=threshold,
                                 symbols_type=symbols_type,
                                 symbols_tuple=symbol_set)[1:] # excluding bias
 
    total_time = round(time.time() - start_time, 2)

    # Statistics 
    print "\n"
    print "Statistics: "
    print "Time taken: {}s".format(total_time)
    print
    
    # Pre-rescore test stats
    print "Pre-rescore test stats: "
    print_stats(X_test, y_test, old_scores, 5)
    print_stats(X_test, y_test, old_scores, 15)
    
    # Post-rescore test stats
    print
    print "Post-rescore test data stats:"
    print_stats(X_test, y_test, new_scores, args.threshold)
    print_stats(X_test, y_test, new_scores, args.threshold)
    print
    
    print_new_scores(output_file=output,
                     symbol_set=symbol_set,
                     symbols_type=symbols_type,
                     new_scores=new_scores)

    
if __name__ == "__main__":
    main()

File: src/rescore/lua/rescore.lua
local argparse = require "argparse"
local torch = require "torch"
local nn = require "nn"
local lua_util = require "lua_util"
local rspamd_logger = require "rspamd_logger"
local ucl = require "ucl"
local nninit = require "nninit"

local rescore_utility = require "rescore_utility"

local function weight_to_score(weight, bias, threshold) 
   return weight * threshold / bias
end

local function score_to_weight(weight,  bias, threshold)
   return weight * bias / threshold
end

local function make_dataset_from_logs(logs, all_symbols)
   -- Returns a list of {input, output} for torch SGD train

   local dataset = {}
   
   for temp, log in pairs(logs) do
      input = torch.Tensor(#all_symbols)
      output = torch.Tensor(1)
      log = lua_util.rspamd_str_split(log, " ")

      if log[1] == "SPAM" then
	 output[1] = 1
      else
	 output[1] = 0
      end
      
      local symbols_set = {}

      for i=4,#log do
	 symbols_set[log[i]] = true
      end

      for index, symbol in pairs(all_symbols) do
	 if symbols_set[symbol] then
	    input[index] = 1
	 else
	    input[index] = 0
	 end	 
      end

      dataset[#dataset + 1] = {input, output}

   end

   function dataset:size()
      return #dataset
   end
   
   return dataset
end

local function init_weights(all_symbols, original_symbol_scores)

   local weights = torch.Tensor(#all_symbols)
   local size = weights:size()[1]
   
   local mean = 0
   
   for i, symbol in pairs(all_symbols) do
      local score = original_symbol_scores[symbol]
      if not score then score = 0 end
      weights[i] = score
      mean = mean + score
   end
--[[
   mean = mean / size

   local dev = 0
   for i=1,size do
      dev = dev + (weights[i] - mean) * (weights[i] - mean)
   end

   dev = math.sqrt(dev / size)
   
   for i=1,size do
      weights[i] = (weights[i] - mean) / dev
   end
]]
   return weights   
end

local function shuffle(logs)

   local size = #logs
   for i = size, 1, -1 do
      local rand = math.random(size)
      logs[i], logs[rand] = logs[rand], logs[i]
   end

end

local function split_logs(logs, split_percent)

   if not split_percent then
      split_percent = 60
   end

   local split_index = #logs * split_percent / 100

   local test_logs = {}
   local train_logs = {}

   for i=1,split_index do
      train_logs[#train_logs + 1] = logs[i]
   end   
   
   for i=split_index + 1, #logs do
      test_logs[#test_logs + 1] = logs[i]
   end

   return train_logs, test_logs
end

local function stitch_new_scores(all_symbols, new_scores)

   local new_symbol_scores = {}

   for idx, symbol in pairs(all_symbols) do
      new_symbol_scores[symbol] = new_scores[idx]
   end

   return new_symbol_scores
end


local function update_logs(logs, symbol_scores)

   for i, log in ipairs(logs) do
      log = lua_util.rspamd_str_split(log, " ")
      local score = 0
      for i=4,#log do
	 log[i] = log[i]:gsub("%s+", "")
	 score = score + (symbol_scores[log[i]] or 0)
      end
      log[2] = rescore_utility.round(score, 2)

      logs[i] = table.concat(log, " ")
   end

   return logs
end

local function write_scores(new_symbol_scores, file_path)
   
   local file = assert(io.open(file_path, "w"))

   local new_scores_json = ucl.to_format(new_symbol_scores, "json")

   file:write(new_scores_json)
   
   file:close()
end

local function print_score_diff(new_symbol_scores, original_symbol_scores)

   print(string.format("%-35s %-10s %-10s", "SYMBOL", "OLD_SCORE", "NEW_SCORE"))

   for symbol, new_score in pairs(new_symbol_scores) do
      print(string.format("%-35s %-10s %-10s",
			  symbol,
			  original_symbol_scores[symbol] or 0,
			  rescore_utility.round(new_score, 2)))
   end

   print "\nClass changes \n"
   for symbol, new_score in pairs(new_symbol_scores) do
      if original_symbol_scores[symbol] ~= nil then
	 if (original_symbol_scores[symbol] > 0 and new_score < 0) or
	 (original_symbol_scores[symbol] < 0 and new_score > 0) then
	       print(string.format("%-35s %-10s %-10s",
				   symbol,
				   original_symbol_scores[symbol] or 0,
				   rescore_utility.round(new_score, 2)))
	 end
      end

   end
      
end

local function calculate_fscore_from_weights(logs, all_symbols, weights, bias, threshold)

   local scale_factor = threshold / bias
   local new_symbol_scores = weights:clone()

   new_symbol_scores:apply( function(wt) wt = wt * scale_factor end )
   new_symbol_scores = stitch_new_scores(all_symbols, new_symbol_scores)

   logs = update_logs(logs, new_symbol_scores)

   local file_stats, _ = rescore_utility.generate_statistics_from_logs(logs, threshold)
   
   return file_stats.fscore
end

local function print_stats(logs, threshold)

   local file_stats, _ = rescore_utility.generate_statistics_from_logs(logs, threshold)

   local file_stat_format = [[
F-score: %.2f
False positive rate: %.2f %%
False negative rate: %.2f %%
Overall accuracy: %.2f %%
]]

   io.write("\nStatistics at threshold: " .. threshold .. "\n")
   
   io.write(string.format(file_stat_format,
			  file_stats.fscore,
			  file_stats.false_positive_rate,
			  file_stats.false_negative_rate,
			  file_stats.overall_accuracy))

end

local parser = argparse() {
   name = "rescore",
   description = "Rescore symbols using perceptron"
}

parser:option("-i --iters", "Number of iterations", 500, tonumber)
parser:option("-l --logdir", "Path to log files")
parser:option("-r --rate", "Learning rate", 1, tonumber)
parser:option("-t --threshold", "Set spam threshold", 15, tonumber)
parser:option("-o --output", "Write new scores to file in json")
parser:flag("--diff", "Print score diff")

parser:mutex(
   parser:flag("--tanh", "Use tanh as activation function instead of sigmoid"),
   parser:flag("--relu", "Use ReLU as activation function instead of sigmoid"),
   parser:flag("--leakyrelu", "Use leaky ReLU as activation function instead of sigmoid")
)


local params = parser:parse()

local logs = rescore_utility.get_all_logs(params.logdir)
local all_symbols = rescore_utility.get_all_symbols(logs)
local original_symbol_scores = rescore_utility.get_all_symbol_scores()

shuffle(logs)

local train_logs, test_logs = split_logs(logs, 70)
local cv_logs, test_logs = split_logs(test_logs, 50)

local dataset = make_dataset_from_logs(train_logs, all_symbols)

local learning_rates = {0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 2.5, 5, 7.5, 10}
local penalty_weights = {0, 0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 3, 5, 10, 15, 20, 25, 50, 75, 100}

-- Start of perceptron training

local input_size = #all_symbols
local linear_module = nn.Linear(input_size, 1)
   
local perceptron = nn.Sequential()
perceptron:add(linear_module)

local activation

if params.tanh then
   activation = nn.Tanh()
elseif params.relu then
   activation = nn.ReLU()
elseif params.leakyrelu then
   activation = nn.LeakyReLU()
else
   activation = nn.Sigmoid()
end

perceptron:add(activation)

local criterion = nn.MSECriterion()
criterion.sizeAverage = false

local best_fscore = -math.huge
local best_weights = linear_module.weight[1]:clone()
local best_weights_bias = linear_module.bias[1]
local l1_weight = 0
local l2_weight = 0

trainer = nn.StochasticGradient(perceptron, criterion)
trainer.maxIteration = params.iters
trainer.verbose = false

local function regularize_parameters(network, l1_weight, l2_weight)

   local parameters, _ = network:parameters()
   for i = 1, table.getn(parameters) do
      local update = torch.clamp(parameters[i], -l1_weight, l1_weight)
      update:add(parameters[i]:mul(-l2_weight))
      parameters[i]:csub(update)      
   end
end

trainer.hookIteration = function(self, iteration, error)

   if iteration == trainer.maxIteration then

      fscore = calculate_fscore_from_weights(cv_logs,
					     all_symbols,
					     linear_module.weight[1],
					     linear_module.bias[1],
					     params.threshold)
      
      print("Cross-validation fscore: " .. fscore)
      
      if best_fscore < fscore then
	 best_fscore = fscore
	 best_weights = linear_module.weight[1]:clone()
	 best_weights_bias = linear_module.bias[1]
      end
   end
end

for _, learning_rate in pairs(learning_rates) do
   for _, weight in pairs(penalty_weights) do

      trainer.weightDecay = weight
      print("Learning with learning_rate: " .. learning_rate .. " | l2_weight: " .. weight)
   
      linear_module.weight[1] = init_weights(all_symbols, original_symbol_scores)

      trainer.learningRate = learning_rate
      trainer:train(dataset)
   
      print()
   end   
end

-- End perceptron training

local scale_factor = params.threshold / best_weights_bias

local new_symbol_scores = best_weights

new_symbol_scores:apply( function(wt) wt = wt * scale_factor end )

new_symbol_scores = stitch_new_scores(all_symbols, new_symbol_scores)

if params.output then
   write_scores(new_symbol_scores, params.output)
end

if params.diff then
   print_score_diff(new_symbol_scores, original_symbol_scores)
end


-- Pre-rescore test stats
print("\n\nPre-rescore test stats\n")
test_logs = update_logs(test_logs, original_symbol_scores)
print_stats(test_logs, 15)

-- Post-rescore test stats
test_logs = update_logs(test_logs, new_symbol_scores)
print("\n\nPost-rescore test stats\n")
print_stats(test_logs, 15)

